File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/user/UserPreferences.java
package top.likoslupus.cellulosesz.api.user;

import org.jspecify.annotations.Nullable;

import java.util.UUID;

public record UserPreferences(
        boolean privateMessages,
        boolean payments,
        boolean teleportRequests,
        boolean teleportAutoAccept,
        boolean confirmLargePayments,
        boolean confirmInventoryClears,
        boolean replyToLastRecipient,
        boolean powerToolsEnabled,
        boolean socialSpy,
        @Nullable UUID incomingReplyTarget,
        @Nullable UUID outgoingReplyTarget
) {

    public static UserPreferences defaults() {
        return new UserPreferences(
                true,
                true,
                true,
                false,
                true,
                true,
                false,
                true,
                false,
                null,
                null
        );
    }

    public UserPreferences withPrivateMessages(boolean value) {
        return copy(value, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    private static UserPreferences copy(
            boolean privateMessages,
            boolean payments,
            boolean teleportRequests,
            boolean teleportAutoAccept,
            boolean confirmLargePayments,
            boolean confirmInventoryClears,
            boolean replyToLastRecipient,
            boolean powerToolsEnabled,
            boolean socialSpy,
            @Nullable UUID incomingReplyTarget,
            @Nullable UUID outgoingReplyTarget
    ) {
        return new UserPreferences(
                privateMessages,
                payments,
                teleportRequests,
                teleportAutoAccept,
                confirmLargePayments,
                confirmInventoryClears,
                replyToLastRecipient,
                powerToolsEnabled,
                socialSpy,
                incomingReplyTarget,
                outgoingReplyTarget
        );
    }

    public UserPreferences withPayments(boolean value) {
        return copy(privateMessages, value, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withTeleportRequests(boolean value) {
        return copy(privateMessages, payments, value, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withTeleportAutoAccept(boolean value) {
        return copy(privateMessages, payments, teleportRequests, value, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withConfirmLargePayments(boolean value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, value, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withConfirmInventoryClears(boolean value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, value, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withReplyToLastRecipient(boolean value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, value, powerToolsEnabled, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withPowerToolsEnabled(boolean value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, value, socialSpy, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withSocialSpy(boolean value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, value, incomingReplyTarget, outgoingReplyTarget);
    }

    public UserPreferences withIncomingReplyTarget(@Nullable UUID value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, value, outgoingReplyTarget);
    }

    public UserPreferences withOutgoingReplyTarget(@Nullable UUID value) {
        return copy(privateMessages, payments, teleportRequests, teleportAutoAccept, confirmLargePayments, confirmInventoryClears, replyToLastRecipient, powerToolsEnabled, socialSpy, incomingReplyTarget, value);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/user/UserPreferences.kt
package top.likoslupus.cellulosesz.api.user

import java.util.*

@JvmRecord
public data class UserPreferences(
    public val privateMessages: Boolean,
    public val payments: Boolean,
    public val teleportRequests: Boolean,
    public val teleportAutoAccept: Boolean,
    public val confirmLargePayments: Boolean,
    public val confirmInventoryClears: Boolean,
    public val replyToLastRecipient: Boolean,
    public val powerToolsEnabled: Boolean,
    public val socialSpy: Boolean,
    public val incomingReplyTarget: UUID?,
    public val outgoingReplyTarget: UUID?
) {

    public fun withPrivateMessages(value: Boolean): UserPreferences =
        copy(privateMessages = value)

    public fun withPayments(value: Boolean): UserPreferences =
        copy(payments = value)

    public fun withTeleportRequests(value: Boolean): UserPreferences =
        copy(teleportRequests = value)

    public fun withTeleportAutoAccept(value: Boolean): UserPreferences =
        copy(teleportAutoAccept = value)

    public fun withConfirmLargePayments(value: Boolean): UserPreferences =
        copy(confirmLargePayments = value)

    public fun withConfirmInventoryClears(value: Boolean): UserPreferences =
        copy(confirmInventoryClears = value)

    public fun withReplyToLastRecipient(value: Boolean): UserPreferences =
        copy(replyToLastRecipient = value)

    public fun withPowerToolsEnabled(value: Boolean): UserPreferences =
        copy(powerToolsEnabled = value)

    public fun withSocialSpy(value: Boolean): UserPreferences =
        copy(socialSpy = value)

    public fun withIncomingReplyTarget(value: UUID?): UserPreferences =
        copy(incomingReplyTarget = value)

    public fun withOutgoingReplyTarget(value: UUID?): UserPreferences =
        copy(outgoingReplyTarget = value)

    public companion object {

        @JvmStatic
        public fun defaults(): UserPreferences =
            UserPreferences(
                privateMessages = true,
                payments = true,
                teleportRequests = true,
                teleportAutoAccept = false,
                confirmLargePayments = true,
                confirmInventoryClears = true,
                replyToLastRecipient = false,
                powerToolsEnabled = true,
                socialSpy = false,
                incomingReplyTarget = null,
                outgoingReplyTarget = null
            )

    }

}
